### PR TITLE
Add state regeneration queue to limit concurrent regenerations

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
To avoid excessive memory use and avoid duplicate block processing, introduce a `StateRegenerationQueue` which schedules regeneration of states ensuring that a requested state is only generated once (multiple requests for the same state are coalesced).  Also identifies when a better starting state is already being generated and "rebases" latest requests to use that state instead of reprocessing the same blocks (ie if we are asked to process A,B,C,D then B,C,D,E just wait for D to be produced then apply E on top of that result rather than reprocessing B, C and D).

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateGenerationQueue.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+      final HashTree treeFromAncestor =
+          tree.withRoot(newBaseRoot).block(newBaseBlockAndState.getBlock()).build();
+      return new RegenerationTask(
+          blockRoot, treeFromAncestor, newBaseBlockAndState, blockProvider, cacheHandler);
+    }
+
+    public SafeFuture<SignedBlockAndState> regenerate() {
+      final StateGenerator stateGenerator =
+          StateGenerator.create(tree, baseBlockAndState, blockProvider);
+      return stateGenerator.regenerateStateForBlock(blockRoot).thenPeek(cacheHandler);
```
Comment (mbaxter) [inline ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateGenerationQueue.java:201]: Something to potentially experiment with in a follow-up PR:  we could use [`AsyncChainStateGenerator.generateStates(targetRoot, stateHandler)`](https://github.com/PegaSysEng/teku/blob/846737706485cd0a79ec1e6b1a3d56b7386e0e93/ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/AsyncChainStateGenerator.java#L85) to expose an API that gives us access to all intermediate states as well:  `StateGenerator.regenerateStatesForBlock(targetRoot, stateHandler)`.  We could then dedupe requests across the entire chain being processed.

File: ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/StateGenerationQueueTest.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    public void assertRegeneratedWithoutRebase() {
+      assertThat(rebasedTo).isEmpty();
+      assertThat(regenerated).isTrue();
+    }
+
+    public void assertNotRegenerated() {
+      assertThat(regenerated).isFalse();
+    }
+
+    public void assertRegneratedAfterRebase(final SignedBlockAndState newBaseState) {
```
Comment (mbaxter) [inline ethereum/core/src/test/java/tech/pegasys/teku/core/stategenerator/StateGenerationQueueTest.java]:
```suggestion
    public void assertRegeneratedAfterRebase(final SignedBlockAndState newBaseState) {
```

**Review by mbaxter — APPROVED** [review]
LGTM

File: ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateGenerationQueue.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateGenerationQueue.java:201]: Interesting idea, the one concern I'd have is making sure we didn't consume too much memory in the process.

**Pull request merged**